Composition

rgbkrk opened the pull request:
I've done a read-only view, now it's time to think about how people might want to work with the _live_ components. I'm trying out some [compound components](https://github.com/rgbkrk/react-workshop/blob/master/subjects/CompoundComponents/lecture.js) ideas that I practiced in the react workshop.

rgbkrk commented:
Current status:

![screen shot 2017-11-08 at 11 30 46 am](https://user-images.githubusercontent.com/836375/32570206-4e3ddcf8-c478-11e7-94c9-ee47dcff2422.png)

rgbkrk commented:
Let's say you want the prompts to be on the right instead.

You'd change the code like this (moving the prompt to be after the `<Editor />`):

```diff
diff --git a/components/composable/notebook.js b/components/composable/notebook.js
index dbd7ac4..016546e 100644
--- a/components/composable/notebook.js
+++ b/components/composable/notebook.js
@@ -364,11 +364,6 @@ export const Example = () => (
           {cell.type === "code" ||
           (cell.type === "markdown" && cell.mode === "edit") ? (
             <Input>
-              {cell.type === "markdown" ? (
-                <PromptBuffer />
-              ) : (
-                <Prompt running={cell.running} queued={cell.queued} />
-              )}
               <Editor>
                 <SyntaxHighlighter
                   style={idea}
@@ -383,6 +378,11 @@ export const Example = () => (
                   {cell.code}
                 </SyntaxHighlighter>
               </Editor>
+              {cell.type === "markdown" ? (
+                <PromptBuffer />
+              ) : (
+                <Prompt running={cell.running} queued={cell.queued} />
+              )}
             </Input>
           ) : null}
           <Outputs>
```

Voila!

![screen shot 2017-11-08 at 12 09 33 pm](https://user-images.githubusercontent.com/836375/32571897-d76a4b74-c47d-11e7-82cb-b95c7a6f724b.png)

Which would also mean you could make markdown cells be RTL a little more easily too (set `textAlign` to right, move those prompts, etc.).

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+
+type OutputsProps = {
+  children: React$Node,
+  className?: string
```
Should the output prompt have a prop that indicates it's hidden/non-hidden status as well?

That way in the render we can do something like:

```
if (this.props.children && this.props.visible) { ... } else { return null };
```

Not sure if this is the right kind of state to have in the prop though.

rgbkrk replied:
I don't have a good heuristic here, I guess we'd have to compare whether we want it to be on the outside or not (also note that I used `hidden` as the prop since I'd want the default to be visible with an explicit property when not):

```
<Cell>
  <Input />
  <Outputs hidden />
</Cell>
```

vs.

```
<Cell>
  <Input />
  { visible ? <Outputs /> : null }
</Cell>
```

People can always do the second one whether or not we have the `hidden` interface. I have to admit I prefer the first, even though I'm not sure what benefit it would have inside the component

captainsafia replied:
> also note that I used hidden as the prop since I'd want the default to be visible with an explicit property when not

That makes sense!

I agree with you. IMO, having it within the component feels like a more intuitive option for the end user.
1. It reduces the amount of code that I have to write.
2. It's much easier to see that `hidden` is associated with the `Output`.

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+import { idea } from "react-syntax-highlighter/dist/styles";
+
+export const Passthrough = (props: { children?: React$Node }) =>
+  props.children || null;
```
I might be missing something but why is this needed?

rgbkrk replied:
This was here for prototyping so that I could make new elements by just writing:

```js
const CellOptions = Passthrough
const CellCreatorToolbar = Passthrough
```

Which let me think on the interface inside of the compound components without it actually affecting anything yet:

```js
<Notebook>
  <CellCreatorToolbar />
  <Cell>
    <CellOptions />
    <Input>
      <Prompt />
      <Editor />
    </Input>
    <Outputs />
  </Cell>
  <CellCreatorToolbar />
</Notebook>
```

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+      } else if (child.type === Editor) {
+        return (
+          <div className="input">
+            {React.cloneElement(child, { className: "input" })}
```
Why are you passing the `className` prop to the `Editor` element if it is already set as the default?

rgbkrk replied:
I was running into CSS issues, looks like I can delete the wrapper div now and just set it on the `child` (which was necessary for the `ReactHighlighter` to look right).

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+}
+
+type InputProps = {
+  children: React$Node
```
In a similar fashion to above, should the `Input` component also contain a `visible` prop?

rgbkrk replied:
You're definitely making me think so. Thanks for thinking about other parts of our current implementation & usage I should consider here.

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+};
+
+// Totally fake component for Markdown consistency
+class PromptBuffer extends React.Component<*> {
```
MAARRRRKKKDOWWWWNNNN!

rgbkrk replied:
YEAHHHHHHHHHHHHHH

captainsafia commented on components/composable/notebook.js:
```diff
@@ -0,0 +1,399 @@
+    if (!child) {
+      return null;
+    }
+    if (child.type === Prompt || child.type === PromptBuffer) {
```
Nitpick: but this might be a good place for a `switch` statement.

rgbkrk commented:
Ok, I've addressed feedback for the current components and made a new little demo notebook + a demo dashboard:

![screen shot 2017-11-08 at 3 40 54 pm](https://user-images.githubusercontent.com/836375/32580683-46572116-c49b-11e7-8789-e71b8df3279a.png)

rgbkrk commented:
![screen shot 2017-11-08 at 3 51 41 pm](https://user-images.githubusercontent.com/836375/32580958-bd044edc-c49c-11e7-9054-f087ac4e48fa.png)

rgbkrk commented:
OK I'm pretty happy with the state of these components, I think it's time for me to try to bring these into `@nteract/core`.

rgbkrk commented:
You were soooooo right @captainsafia, as a result of letting the container element determine hidden (rather than making a nested ternary that returns `null` some of the time), the code is soooo clean. I'm currently reworking notebook-preview and this is how the current code is looking:

```jsx
    return (
      <Notebook>
        {cellOrder.map(cellID => {
          const cell = cellMap.get(cellID);
          const sourceHidden =
            allSourceHidden ||
            cell.getIn(["metadata", "inputHidden"]) ||
            cell.getIn(["metadata", "hide_input"]);

          const outputHidden = cell.get("outputs").size === 0;

          switch (cell.get("cell_type")) {
            case "code":
              return (
                <Cell key={cellID}>
                  <Input hidden={sourceHidden}>
                    <Prompt />
                    <Editor language={language}>{cell.get("source")}</Editor>
                  </Input>
                  <Outputs hidden={outputHidden}>
                    <Display outputs={cell.get("outputs").toJS()} />
                  </Outputs>
                </Cell>
              );

```
